apps/prairielearn/src/pages/workspace/workspace.html.ts

```diff
@@ -33,6 +34,21 @@ export function Workspace({
         ${HeadContents({ resLocals, pageNote, pageTitle })}
         <link href="${assetPath('stylesheets/workspace.css')}" rel="stylesheet" />
         ${compiledScriptTag('workspaceClient.ts')}
+        ${resLocals.assessment &&
+        resLocals.assessment.type === 'Exam' &&
+        resLocals.assessment_instance_remaining_ms
+          ? html`${compiledScriptTag('examTimeLimitCountdown.ts')}
+            ${EncodedData(
+              {
+                serverRemainingMS: resLocals.assessment_instance_remaining_ms,
+                serverTimeLimitMS: resLocals.assessment_instance_time_limit_ms,
+                serverUpdateURL: `${resLocals.urlPrefix}/assessment_instance/${resLocals.assessment_instance.id}/time_remaining`,
```

Comment:
`resLocals.urlPrefix` isn't set to any particularly helpful value on the workspace routes, so this will end up pointing to a route that doesn't exist (it 404s). You can reproduce this for yourself: change the refresh interval here from `60000` to `1000` (once per second):

https://github.com/PrairieLearn/PrairieLearn/blob/b5869473bebf56c952d6c1a7e710556505fb1780/apps/prairielearn/assets/scripts/lib/countdown.ts#L74

Now observe how the page reloads once per second because of this:

https://github.com/PrairieLearn/PrairieLearn/blob/b5869473bebf56c952d6c1a7e710556505fb1780/apps/prairielearn/assets/scripts/examTimeLimitCountdown.ts#L39

You should manually construct `serverUpdateURL` to be an appropriate value; see the timers on other pages to get a sense of what that route should be.

@mwest1066 @jonatanschroeder I was a little surprised that we unconditionally refresh if the server update fails. I had the same question the last time this code came up: https://github.com/PrairieLearn/PrairieLearn/pull/7967#discussion_r1245501865. _Why_ do we do this - is this to pick up a change in access rules or something? I'm a little concerned that this has the potential to destroy student work on this page.

Reply:
Wow, I can't believe that I missed that when testing the timers. I must've always used 1 minute as exam time for testing so it never actually triggered an update.

The final update (when the timeout happens) can still fail and trigger a reload by the way because the time_remaining API stops working the moment the assessment closes, and it produces a 403 error afterwards. I think this only really becomes an issue if a student has multiple pages with countdowns running in different tabs, and it's not really a problem because at that point any unsaved work wouldn't be graded anyway. However, it's still not the cleanest way to handle this, and it might give students an unnecessary moment of panic if multiple pages in their browser suddenly refresh. 

One way to fix this would be to adjust the access rules so that the time_remaining API also works for closed assessments, or another would be to not do a hard reload as you mentioned earlier.

Reply:
> @mwest1066 @jonatanschroeder I was a little surprised that we unconditionally refresh if the server update fails. I had the same question the last time this code came up: [#7967 (comment)](https://github.com/PrairieLearn/PrairieLearn/pull/7967#discussion_r1245501865). _Why_ do we do this - is this to pick up a change in access rules or something? I'm a little concerned that this has the potential to destroy student work on this page.

Seems like this was added in #4842, though I can't find any documentation as to why we do that. My guess is that what @nritschel says makes sense: we're trying to capture a 403, which should cause the page to become unavailable, even if temporarily (e.g., if we temporarily remove access on PrairieTest). I agree that a different approach may be better.

Reply:
Discussion on the reload case was moved to #10415, to be handled in a future PR.